Source/DafnyCore/Options/TranslationRecord.cs

```diff
@@ -100,16 +117,44 @@ public class TranslationRecord {
     return null;
   }
 
-  public void Merge(TranslationRecord other) {
-    // TODO: check versions
-    
-    // TODO: This will error if any modules overlap, which is what we want,
-    // but we can do much better in terms of error messages.
+  private void Merge(ErrorReporter reporter, TranslationRecord other, string filePath, IToken origin) {
+    // Assume both this and other have been Validate()-d already.
+
+    var duplicateModules = OptionsByModule
+      .Union(other.OptionsByModule)
+      .GroupBy(p => p.Key)
+      .Where(g => g.Count() > 1)
+      .Select(g => g.Key)
+      .ToList();
+    if (duplicateModules.Any()) {
+      var messagePrefix = $"cannot load {filePath}";
+      foreach (var duplicateModule in duplicateModules) {
+        reporter.Error(MessageSource.Project, origin,
+          $"{messagePrefix}: module {duplicateModule} already appears in another translation record");
+      }
+      return;
+    }
+
     OptionsByModule = OptionsByModule.Union(other.OptionsByModule).ToDictionary(p => p.Key, p => p.Value);
   }
-  
+
+  public static void ReadValidateAndMerge(Program dafnyProgram, string filePath, IToken origin) {
```

Comment:
Would you mind replacing the `Program` argument with `ErrorReporter` (which also contains an Options) and `CompilationData`, to make it easier to see from the caller what is being used?

Reply:
I can't quite do that because `Validate` depends on seeing the set of modules to compile as well, to validate that they don't appear as previously compiled in translation records as well. 

I could pass in `ErrorReporter`, `CompilationData`, and the set of compile modules if you feel strongly about it. I would almost prefer to pass in `Options` separately as well since it's a key parameter that feels weird to access through the error reporter to me.

Reply:
Ah, never mind I'm fine with the current code. Maybe rename `dafnyProgram` to `program`, and order the methods from high to low level, so `ReadValidateAndMerge` above `Validate`